crate_universe/extension.bzl

```diff
@@ -0,0 +1,218 @@
+    # an input to the rule, this would invalidate the repo rule,
+    # requiring it to be rerun on every invocation.
+    # To solve this, we allow it to touch a copy of the lock file,
+    # rather than the original.
+    cargo_lockfile_copy = tag_path.get_child("copy/Cargo.lock")
+    module_ctx.file(
+        cargo_lockfile_copy,
+        module_ctx.read(cargo_lockfile),
+    )
+    cargo_bazel([
+        "generate",
+        "--cargo-lockfile",
+        cargo_lockfile_copy,
+        "--config",
+        config_file,
+        "--splicing-manifest",
+        splicing_manifest,
+        "--repository-dir",
+        tag_path,
+        "--metadata",
+        splicing_output_dir.get_child("metadata.json"),
+        "--repin",
+        "--lockfile",
+        lockfile_path,
+    ])
+
+    crates_dir = tag_path.get_child(repo_name)
+    _generate_repo(
+        name = repo_name,
+        contents = {
+            "BUILD.bazel": module_ctx.read(crates_dir.get_child("BUILD.bazel")),
+        },
+    )
+
+    contents = json.decode(module_ctx.read(lockfile_path))
+
+    for crate in contents["crates"].values():
+        repo = crate["repository"]
+        if repo == None:
+            continue
+        name = crate["name"]
+        version = crate["version"]
+
+        # "+" isn't valid in a repo name.
+        crate_repo_name = "{repo_name}__{name}-{version}".format(
+            repo_name = repo_name,
+            name = name,
+            version = version.replace("+", "-"),
+        )
+
+        build_file_content = module_ctx.read(crates_dir.get_child("BUILD.%s-%s.bazel" % (name, version)))
+        if "Http" in repo:
+            # Replicates functionality in repo_http.j2.
+            repo = repo["Http"]
+            http_archive(
+                name = crate_repo_name,
+                patch_args = repo.get("patch_args", None),
+                patch_tool = repo.get("patch_tool", None),
+                patches = repo.get("patches", None),
+                remote_patch_strip = 1,
+                sha256 = repo["sha256"],
+                type = "tar.gz",
+                urls = [repo["url"]],
+                strip_prefix = "%s-%s" % (crate["name"], crate["version"]),
+                build_file_content = build_file_content,
+            )
+        elif "Git" in repo:
+            # Replicates functionality in repo_git.j2
+            repo = repo["Git"]
+            new_git_repository(
+                name = crate_repo_name,
+                init_submodules = True,
+                patch_args = repo.get("patch_args", None),
+                patch_tool = repo.get("patch_tool", None),
+                patches = repo.get("patches", None),
+                shallow_since = repo.get("shallow_since", None),
+                remote = repo["remote"],
+                build_file_content = build_file_content,
+                strip_prefix = repo.get("strip_prefix", None),
+                **repo["commitish"]
```

Comment:
Unfortunately this is not working for me when trying to use this branch (at current commit `dd7e67f72a6bc3ef852fe6a424b37e0ebbcaeadc`) to build my own repo, where I need to pull some of my crates via git from non-public github repos (`my-crate = { git = "ssh://git@github.com/...", rev = "..." }`.

The problem is that the generated `cargo-bazel-lock.json` file has keys in the `commitish` object which are capitalized: `Branch`, `Tag`, or `Rev`:

```
    "my-crate ...": {
      "name": "my-crate",
      "version": "...",
      "repository": {
        "Git": {
          "remote": "ssh://git@github.com/...",
          "commitish": {
            "Rev": "..."
          }
        }
      },
 ```
When expanded into the arguments for `new_git_repository` this fails because of course `new_git_repository` needs them lower case. This results in an error like this:

```
$ bazel query @crates//:\*
ERROR: <builtin>: @rules_rust~override//crate_universe:rules_rust~override~crate~crates__my-crate...: no such attribute 'Rev' in 'git_repository' rule
ERROR: Traceback (most recent call last):
        File ".../external/rules_rust~override/crate_universe/extension.bzl", line 194, column 37, in _crate_impl
                _generate_hub_and_spokes(module_ctx, cargo_bazel, mod_path, cfg)
        File ".../external/rules_rust~override/crate_universe/extension.bzl", line 174, column 31, in _generate_hub_and_spokes
                new_git_repository(
Error in repository_rule: failed to instantiate 'git_repository' from this module extension
ERROR: Target parsing failed due to unexpected exception: error evaluating module extension crate in @rules_rust~override//crate_universe:extension.bzl
```

Reply:
Also noting that `cargo-bazel-lock.json` uses `Rev`, but `new_git_repository` uses `commit`. I am finding happiness 
with this line altered to
```
**{"commit" if k == "Rev" else k.lower(): v for k, v in repo["commitish"].items()}
```

Reply:
Should be fixed, though I don't have a git-based repo to test it myself. Please try it yourself to confirm.